designs/delay-node-enhancements/delay-node-enhancements.md

```diff
@@ -0,0 +1,83 @@
+---
+state: draft
+---
+
+# delay node enhancements
+## Summary
+ The enhancements for delay node change message queue handling and provide logging function.
+
+## Author
+ @kazuhitoyokoi
+
+## Details
+### Use cases
+#### 1. Incoming message control
+ Because Node-RED accepts all incoming message from input nodes like mqtt-in node, Node.js process sometimes reaches the upper limit of memory usage.
+To avoid the risk of message queue overflow, rate limit functionality in delay node will be useful but Node-RED users cannot customize queue handling.
+For example, the current delay node cannot simultaneously use the drop option and queue to process burst incoming messages.
+
+#### 2. Changing properties on demand
+ Currently, the delay node supports msg.delay to change property value on demand.
+If the delay node additionally supports msg.rate to set rate value and msg.queueLength to change queue size, users can create custom delay node using subflow to wrap delay node.
+
+#### 3. Logging and error handling
+ Because current drop option in rate limit mode doesn't record dropped messages, Node-RED users aren't aware of the lost messages.
+Therefore, it will be useful for Node-RED users to track background behaviors if the delay node outputs the number of dropped messages to the log stream.
+Additionally, Node-RED users can add their error handling in their flow if the delay node throws the dropped messages to the catch node when the queue reaches the defined limit.
+
+### Options in setting.js which the delay node uses
+- nodeMessageBufferMaxLength
+
+  The enhanced delay node supports nodeMessageBufferMaxLength as the default buffer length.
+  The buffer length needs to be greater than queue length defined in msg.queueLength or node property UI.
+
+- customQueueLength (default: false)
+
+  When the customQueueLength value is true, the node property UI of the enhanced delay node shows customQueueLength as the text input.
+  The default value is 0 because the current implementation doesn't use the queue.
+ 
+  ![nodepropertyui.png](nodepropertyui.png)
+
+- throwDroppedMessages (default: false)
+
+  Other nodes which support nodeMessageBufferMaxLength setting throws an empty message to the catch node when they reach the limit of the buffer length.
+  As the default, the behavior of enhanced delay node will be the same as existing other nodes.
+  If the throwDroppedMessages in setting.js is true, the enhanced delay node will throw dropped messages to the catch node. If the Node-RED supports this handling, Node-RED users are able to write their error handling in the flow.
+
+### Logging
+- The number of dropped messages
+  - debug level: Node-RED outputs the node ID and the number of dropped messages to the log stream every 15 seconds.
```

Comment:
but only if there are any - don't want unnecessary logging for nothing dropped. Is this number dropped in the 15 second interval or total ?

Reply:
The logging function outputs number when dropped messages exist only. This is interval number in 15 seconds not total.